river/ensemble/adaptive_random_forest.py

```diff
@@ -0,0 +1,824 @@
+    def __init__(self,
+                 index_original: int,
+                 base_model: typing.Union[ARFHoeffdingTreeClassifier, ARFHoeffdingTreeRegressor],
+                 created_on: int,
+                 base_drift_detector: base.DriftDetector,
+                 base_warning_detector: base.DriftDetector,
+                 is_background_learner,
+                 base_metric: typing.Union[MultiClassMetric, RegressionMetric]):
+        self.index_original = index_original
+        self.base_model = base_model
+        self.model = copy.deepcopy(base_model)
+        self.created_on = created_on
+        self.is_background_learner = is_background_learner
+        self.base_metric = base_metric
+        # Make sure that the metric is not initialized, e.g. when creating background learners.
+        if isinstance(self.base_metric, MultiClassMetric):
+            self.base_metric.cm.reset()
+        self.metric = copy.deepcopy(base_metric)
+
+        self.background_learner = None
+
+        # Drift and warning detection
+        self.base_drift_detector = base_drift_detector  # Drift detector prototype
+        self.base_warning_detector = base_warning_detector  # Warning detector prototype
+
+        self.last_drift_on = 0
+        self.last_warning_on = 0
+        self.n_drifts_detected = 0
+        self.n_warnings_detected = 0
+
+        # Initialize drift and warning detectors
+        # TODO Replace deepcopy with clone
+        if base_drift_detector is not None:
+            self._use_drift_detector = True
+            self.drift_detector = copy.deepcopy(base_drift_detector)  # Actual detector used
+        else:
+            self._use_drift_detector = False
+            self.drift_detector = None
+
+        if base_warning_detector is not None:
+            self._use_background_learner = True
+            self.warning_detector = copy.deepcopy(base_warning_detector)  # Actual detector used
+        else:
+            self._use_background_learner = False
+            self.warning_detector = None
+
+    def reset(self, n_samples_seen):
+        if self._use_background_learner and self.background_learner is not None:
+            # Replace foreground model with background model
+            self.model = self.background_learner.model
+            self.warning_detector = self.background_learner.warning_detector
+            self.drift_detector = self.background_learner.drift_detector
+            self.metric = copy.deepcopy(self.background_learner.base_metric)
+            self.created_on = self.background_learner.created_on
+            self.background_learner = None
+        else:
+            # Reset model
+            self.model = copy.deepcopy(self.base_model)
+            self.metric = copy.deepcopy(self.base_metric)
+            self.created_on = n_samples_seen
+            self.drift_detector = copy.deepcopy(self.base_drift_detector)
+
+    def learn_one(self, x: dict, y: base.typing.ClfTarget, *, sample_weight: int,   # noqa
```

Comment:
Change to `base.typing.Target`

Reply:
Changed in b628832e1062fda4cdbb7ae1c1948d4d6212bd90